plugins/inputs/directory_monitor/directory_monitor.go

```diff
@@ -0,0 +1,468 @@
+
+	// Handle a file read error. We don't halt execution but do document, log, and move the problematic file.
+	if err != nil {
+		monitor.Log.Errorf("Error while reading file: '" + filePath + "'. " + err.Error())
+		monitor.filesDropped.Incr(1)
+		if monitor.UseErrorDirectory {
+			monitor.moveFile(filePath, monitor.ErrorDirectory)
+		}
+		return
+	}
+
+	// File is finished, move it to the 'finished' directory.
+	monitor.moveFile(filePath, monitor.FinishedDirectory)
+	monitor.filesProcessed.Incr(1)
+}
+
+func (monitor *DirectoryMonitor) ingestFile(filePath string) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	parser, err := monitor.parserFunc()
+	if err != nil {
+		return fmt.Errorf("E! Creating parser: %s", err.Error())
+	}
+
+	// Handle gzipped files.
+	var reader io.Reader
+	if filepath.Ext(filePath) == ".gz" {
+		reader, err = gzip.NewReader(file)
+		if err != nil {
+			return err
+		}
+	} else {
+		reader, _ = utfbom.Skip(monitor.decoder.Reader(file))
+	}
+
+	return monitor.parseFile(parser, reader)
+}
+
+func (monitor *DirectoryMonitor) parseFile(parser parsers.Parser, reader io.Reader) error {
+	// JSON files can't be read line-by-line with the current parser.
+	// As a temporary strategy, read them whole, recognizing the potential memory trade-off.
+	if _, isJSONParser := parser.(*json.Parser); isJSONParser {
+		fileContents, err := ioutil.ReadAll(reader)
+		if err != nil {
+			return err
+		}
+
+		metrics, err := parser.Parse(fileContents)
+		if err != nil {
+			return err
+		}
+
+		monitor.sendMetrics(metrics)
+		return nil
+	}
+
+	// Read the file line-by-line and parse with the configured parse method.
+	firstLine := true
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		metrics, err := monitor.parseText(parser, scanner.Text(), firstLine)
+		if err != nil {
+			return err
+		}
+		if firstLine {
+			firstLine = false
+		}
+
+		monitor.sendMetrics(metrics)
+	}
+
+	return nil
+}
+
+func (monitor *DirectoryMonitor) parseText(parser parsers.Parser, line string, firstLine bool) ([]telegraf.Metric, error) {
+	switch parser.(type) {
+	case *csv.Parser:
+		// The CSV parser parses headers in Parse and skips them in ParseLine.
+		if firstLine {
+			return parser.Parse([]byte(line))
+		}
+
+		m, err := parser.ParseLine(line)
+		if err != nil {
+			return nil, err
+		}
+
+		if m != nil {
+			return []telegraf.Metric{m}, nil
+		}
+
+		return []telegraf.Metric{}, nil
+	default:
+		return parser.Parse([]byte(line))
+	}
+}
+
+func (monitor *DirectoryMonitor) sendMetrics(metrics []telegraf.Metric) {
+	// Report the metrics for the file.
+	for _, m := range metrics {
+		// Try writing out metric first without blocking.
+		select {
+		case monitor.sem <- empty{}:
+			monitor.acc.AddTrackingMetricGroup([]telegraf.Metric{m})
+			continue
+		default:
+		}
+
+		// Block until room is available to add metrics.
+		select {
+		case monitor.sem <- empty{}:
```

Comment:
Ideally we need to watch for quit here, unless you intend to block Telegraf shutdown, which might be the case. may require discussion.